tools/auth/webex.go

```diff
@@ -0,0 +1,80 @@
+package auth
+
+import (
+	"context"
+	"encoding/json"
+
+	"github.com/pocketbase/pocketbase/tools/types"
+	"golang.org/x/oauth2"
+)
+
+func init() {
+	Providers[NameWebex] = wrapFactory(NewWebexProvider)
+}
+
+var _ Provider = (*Webex)(nil)
+
+// NameWebex is the unique name of the Webex provider.
+const NameWebex string = "webex"
+
+// Webex allows authentication via Webex OAuth2.
+type Webex struct {
+	BaseProvider
+}
+
+// NewWebexProvider creates a new Webex provider instance with some defaults.
+func NewWebexProvider() *Webex {
+	return &Webex{BaseProvider{
+		ctx:         context.Background(),
+		displayName: "Webex",
+		pkce:        true,
+		scopes:      []string{"spark:people_read"},
+		authURL:     "https://webexapis.com/v1/authorize",
+		tokenURL:    "https://webexapis.com/v1/access_token",
+		userInfoURL: "https://webexapis.com/v1/people/me",
+	}}
+}
+
+// FetchAuthUser returns an AuthUser instance based on the Webex user api.
+//
+// API reference: https://developer.webex.com/docs/api/v1/people/get-my-own-details
+func (p *Webex) FetchAuthUser(token *oauth2.Token) (*AuthUser, error) {
+	data, err := p.FetchRawUserInfo(token)
+	if err != nil {
+		return nil, err
+	}
+
+	rawUser := map[string]any{}
+	if err := json.Unmarshal(data, &rawUser); err != nil {
+		return nil, err
+	}
+
+	extracted := struct {
+		Id          string   `json:"id"`
+		DisplayName string   `json:"displayName"`
+		UserName    string   `json:"userName"`
+		Avatar      string   `json:"avatar"`
+		Emails      []string `json:"emails"`
+	}{}
+	if err := json.Unmarshal(data, &extracted); err != nil {
+		return nil, err
+	}
+
+	user := &AuthUser{
+		Id:           extracted.Id,
+		Name:         extracted.DisplayName,
+		Username:     extracted.UserName,
+		AvatarURL:    extracted.Avatar,
+		RawUser:      rawUser,
+		AccessToken:  token.AccessToken,
+		RefreshToken: token.RefreshToken,
+	}
+
+	if len(extracted.Emails) == 1 {
+		user.Email = extracted.Emails[0]
+	}
```

Comment:
If users can have more than one email, please change the if  condition to `len(extracted.Emails) > 0`.

Also could you confirm that the returned emails are verified?
If they are not verified then we need to skip the assignment because the email is not safe for linking.

Reply:
Yes, users can have more than one email, but the simple `/v1/people` API only uses one email address. I would agree that the documentation isn't really clear on this for the `GET` verb, but for the `POST` and `PUT` verbs it states:
> `emails`
> array[string] **required**
>
> The email addresses of the person. Only one email address is allowed per person.

We do have an SCIM 2.0 compliant API at [`/v1/scim2-user/get-me`](https://developer.webex.com/docs/api/v1/scim2-user/get-me) that would have a more definitive representation of multiple emails, so it might be worth using that instead.
However, a Webex user can only have multiple email addresses if they use an external authentication source. If they signed up independently through webex.com they can only have one email address.

I can confirm that the email is verified.

- Direct sign-ups through webex.com do not get to the stage where they can create a password and complete account setup without first verifying their email address.
- For OAuth authenticated users must use the same email address their OAuth provider returns for their email. So if you sign up to Webex using Google OAuth you will have to use your `@gmail.com` address for you Webex username (or the email you use as your Google account email, if your account doesn't have Gmail/is a Google Workspace account with custom domain, at which point you will have to have verified it with Google previously).
- For enterprise users the email is whatever their authentication source provides. So I would say it is safe to assume it is verified on their end as well.

It is safe to say that if the user ever arrives at the point where they can trigger PocketBase to make this API call there will be a verified email that is returned.